pkgs/net-doc/net/scribblings/server.scrbl

```diff
@@ -0,0 +1,176 @@
+  closes the listener but does not terminate active connections.
+
+  Each client connection is managed by a new custodian, and each call
+  to @racket[handle] occurs in a new thread (also managed by the
+  connection's custodian).  Each handling thread has an associated
+  supervising thread that shuts down the connection's custodian when
+  the handling thread terminates or when the result of
+  @racket[make-timeout-evt] is ready for synchronization.  Breaks are
+  enabled in handling threads if breaks are enabled when
+  @racket[start-server] is called.  Handling threads need not close
+  the provided input and output ports.
+
+  The @racket[make-timeout-evt] procedure is called with the
+  connection-handling thread, the input port and the output port of
+  every new connection, and a boolean to signal if the handling thread
+  has already been sent a break.  When the event it returns is ready
+  for synchronization and if the handling thread is still running, the
+  handling thread is sent a break and @racket[make-timeout-evt] is
+  called again to produce an event that, when ready for
+  synchronization, will cause the connection's custodian to be shut
+  down and, consequently, the handling thread to be killed if it is
+  still running by that time.
+
+  The server keeps track of the number of active connections and
+  pauses accepting new connections once that number reaches
+  @racket[max-concurrent], resuming once the number goes down again.
+
+  The @racket[reuse?] and @racket[max-allow-wait] arguments have the
+  same meaning as in @racket[tcp-listen].  The @racket[listen],
+  @racket[accept] and @racket[close] arguments may be used to run the
+  server on an alternate protocol (such as SSL).  The
+  @racket[listener?] part of the contract indicates that the
+  procedures must all work on the same kind of listener value.  The
+  values returned by @racket[listen] must be synchronizable events
+  that are ready for synchronization when @racket[accept] would not
+  block; the synchronization result of the listener must be a value
+  that can be passed to @racket[accept].
+
+  To listen on and retrieve an ephemeral port, set the @racket[port]
+  argument to @racket[0] and pass a custom @racket[listen] procedure
+  that retrieves the port using @racket[tcp-addresses] in the case of
+  a TCP server.  For example:
+
+  @examples[
+    #:label #f
+    #:eval ev
+    (require net/server
+             racket/tcp)
+    (define ((make-listen-proc port-ch) port backlog reuse? host)
+      (define listener
+        (tcp-listen port backlog reuse? host))
+      (define-values (local-host local-port remote-host remote-port)
+        (tcp-addresses listener #t))
+      (thread (lambda () (channel-put port-ch local-port)))
+      listener)
+    (define port-ch
+      (make-channel))
+    (define stop
+      (start-server
+        #:listen-proc (make-listen-proc port-ch)
+        "127.0.0.1" 0 void))
+    (channel-get port-ch)
+    (stop)
+  ]
```

Comment:
Having `run-server` take `listener` as an argument would simplify this considerably.

Reply:
This was also my major reservation here. The make-listen-proc pattern here seems like it could usefully be a part of the abstraction.

Reply:
My argument here would be that this is a special case. Most users of these procedures are likely not going to be using ephemeral ports and I wouldn't have used them here if I could know for sure that the documentation would build cleanly with specific ports used in the examples. Probably, I should use a log-based evaluator instead for the simple examples and mark this example as a special case.

Reply:
That sounds like a sensible plan; I think we want the examples to be what people would actually do.

Reply:
OK, I've made a separate section for examples and added TCP, TCP+SSL, Unix socket and port examples and dropped the ephemeral port example entirely.